### PR TITLE
[@types/react-leaflet] Update to v2.2

### PR DESCRIPTION
**Pull request by hwu1001**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes:
  - API docs here: https://react-leaflet.js.org/docs/en/components.html
  - Source code here: https://github.com/PaulLeCam/react-leaflet
- [X] Increase the version number in the header if appropriate.
- [X] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

## Notes

Hopefully addresses #27887

Since there are so many changes I wanted to call out some specifically:
* For `FeatureGroup`, `GeoJSON` extends `Path` in react-leaflet. Not sure that I can do the same since the elements used are `Leaflet.FeatureGroup` and `Leaflet.GeoJSON`. In Leaflet it goes `GeoJSON` > `FeatureGroup` > `LayerGroup` > `Layer` > `Evented`.
  * In react-leaflet it goes `GeoJSON`/`FeatureGroup` > `Path` > `MapLayer` > `MapComponent`
  * The react-leaflet inheritance doesn't work in Typescript here since we're using the Leaflet elements. Some options:
  * a) Have `GeoJSON` and `FeatureGroup` extend `MapComponent`
  * b) Have `GeoJSON` and `FeatureGroup` use `Leaflet.Path` for elements
  * c) Could try to have `Path` not specify an element and then alter parent classes of it?
  * d) (Currently using this one) Use the Leaflet inheritance of `GeoJSON` > `FeatureGroup` > `LayerGroup` > `Layer` > `Evented`. The react-leaflet docs seem to support this as well: https://react-leaflet.js.org/docs/en/components.html#featuregroup
    * >Extended `LayerGroup` supporting a Popup child.
* In a few props `leafletContext` is required in the actual library, but it's dynamically injected with `withLeaflet` so I'm not sure that there's a way to not have `leafletContext` optional for the types. If it's marked as required in the types it will throw errors unless it's specifically added to components when declared in JSX.
* For any types in the library that had `property: ?type` I did `type | null | undefined`. I'm not sure if I can just do `property?: type` instead though?
* context.js - looked at this gist for most of it https://gist.github.com/gwillz/9498267353770cbd121fe3aabc5cec68
  * `LeafletConsumer` and `LeafletProvider` - Consumer and Provider went off of this: https://github.com/sw-yx/react-typescript-cheatsheet
  * `withLeaflet` - Omit is from https://stackoverflow.com/questions/48215950/exclude-property-from-type
* `ControlPosition` - This is handled in `Leaflet.ControlOptions`
* `GridLayerOptions` - This is called `GridLayerProps` since `Leaflet.GridLayerOptions` already exists
* `PathOptions` - Using `Leaflet.PathOptions` here
* `DivOverlayOptions` - This is called `DivOverlayProps` since `Leaflet.DivOverlayOptions` already exists
* `LeafletProps` - As far as I can tell this is only in types.js so not including it
* `CircleProps` - this inherits from Object in the library but that doesn't seem right so I haven't added it
* LayersControl.js - need more review on this part. Not sure 100% sure how this should be modeled in TS
* Map - Does this need _ready and _updating properties? I've omitted them for now since they don't seem like they need to be accessible from the types.
* VideoOverlay.js - Not adding this right now since it wasn't previously included

## Testing Notes

Other than running `npm run lint` and `npm run test` I've added the viewport example from the library: https://github.com/PaulLeCam/react-leaflet/blob/master/example/components/viewport.js. Otherwise I added a test to use the context API to create a custom component. I've tested this custom Polygon component in my own React project and confirmed I can do something like the following in a .tsx file:

```jsx
<CustomPolygon positions={[some positions here...]}>
    <Popup>
        Here's some popup text that appears when clicking the polygon.
    </Popup>
</CustomPolygon>
```

### CODE REVIEW

**Review by ZhangYiJiang — COMMENTED** [review]
I've also hacked together my own React Leaflet 2 types for my project when I couldn't find one. Yours looks much more complete. 

For a major version change I *think* the old types should be moved to a `v1` folder. The `onViewportChange` events should probably be backported in that case.

File: types/react-leaflet/index.d.ts [hunk 1]
```diff
@@ -118,234 +118,334 @@ export type LeafletEvents = MapEvents
...
+}
+
+export type ZoomControlProps = Leaflet.Control.ZoomOptions & MapControlProps;
+export class ZoomControl<P extends ZoomControlProps = ZoomControlProps, E extends Leaflet.Control.Zoom = Leaflet.Control.Zoom> extends MapControl<P, E> {
+    createLeafletElement(props: P): E;
+}
+
+// context.js
+export class LeafletConsumer extends React.Component<React.ConsumerProps<LeafletContext>> {}
+export class LeafletProvider extends React.Component<React.ProviderProps<LeafletContext>> {}
```
[the remark this answers is not in the document]
Comment (v1zix) [reply types/react-leaflet/index.d.ts]: I don't think I follow here, are you suggesting to do something like the following?

```ts
export type LeafletConsumer = React.Consumer<LeafletContext>;
export type LeafletProvider = React.Provider<LeafletContext>;
```

If so, I'm not sure how that will work with custom components. For example, with the custom component test I have for the types the `render` method will get a linting error of:
```
'LeafletProvider' only refers to a type, but is being used as a value here.
```

I think `LeafletProvider` has to extend `React.Component`.

[the remark this answers is not in the document]
Comment (ZhangYiJiang) [reply types/react-leaflet/index.d.ts]: No, `LeafletConsumer` and `LeafletProvider` are instances of `React.Consumer` and `React.Provider`. I believe the code for declaring that should be 

```ts
export const LeafletProvider: React.Provider<LeafletContext>;
export const LeafletConsumer: React.Consumer<LeafletContext>;
```

File: types/react-leaflet/index.d.ts [hunk 2]
```diff
@@ -118,234 +118,334 @@ export type LeafletEvents = MapEvents
...
 
-export class Map<P extends MapProps = MapProps, E extends Leaflet.Map = Leaflet.Map> extends MapComponent<P, E> {
-    className?: string;
-    container: HTMLDivElement;
-    getChildContext(): { layerContainer: E, map: E };
+export class Map<P extends MapProps = MapProps, E extends Leaflet.Map = Leaflet.Map> extends MapEvented<P, E> {
+    className: string | null | undefined;
+    contextValue: LeafletContext | null | undefined;
+    container: HTMLDivElement | null | undefined;
+    viewport: Viewport;
```
[the remark this answers is not in the document]
Comment (v1zix) [reply types/react-leaflet/index.d.ts]: I'm not seeing the `viewport` property on the `Map` class as nullable: https://github.com/PaulLeCam/react-leaflet/blob/master/src/Map.js#L103.

`MapProps` has `viewport` as nullable though, which is consistent with what I'm seeing in the library: https://github.com/PaulLeCam/react-leaflet/blob/master/src/Map.js#L95

[the remark this answers is not in the document]
Comment (ZhangYiJiang) [reply types/react-leaflet/index.d.ts]: Yes, I think you're right. I mistook the property on MapProps for Map

File: types/react-leaflet/react-leaflet-tests.tsx [hunk 3]
```diff
@@ -628,6 +633,44 @@ export class VectorLayersExample extends Component<undefined, undefined> {
     }
 }
 
+// viewport.js
+
+const DEFAULT_VIEWPORT = {
+    center: [51.505, -0.09] as [number, number],
```
[the remark this answers is not in the document]
Comment (v1zix) [reply types/react-leaflet/react-leaflet-tests.tsx]: Updated

File: types/react-leaflet/index.d.ts [hunk 4]
```diff
@@ -118,234 +118,334 @@ export type LeafletEvents = MapEvents
...
     createLeafletElement(props: P): E;
     updateLeafletElement(fromProps: P, toProps: P): void;
-    bindContainer(container: HTMLDivElement): void;
+    onViewportChange: () => void;
```
[the remark this answers is not in the document]
Comment (v1zix) [reply types/react-leaflet/index.d.ts]: Good catch, I've updated that now

File: types/react-leaflet/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (v1zix) [reply types/react-leaflet/index.d.ts]: Ah gotcha, I've updated these now

**Comment by v1zix** [comment]
@ZhangYiJiang I've added a `v1/` folder to keep the previous version types. I haven't back-ported the `onViewportChange` events though. Seems like something to do in another PR. Right now I've just taken the code exactly as it is now and put it into the `v1` folder. I only updated `v1/tsconfig.json` so that it has the right file paths listed.

**Comment by danzel** [comment]
Have roughly checked over the types and tested in an internal app, Everything worked.
So :+1: from me. Thanks for doing this!

File: types/react-leaflet/v1/tslint.json [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{
+	"extends": "dtslint/dt.json",
+	"rules": {
+		// TODO
```
Comment (jessetrinity) [inline types/react-leaflet/v1/tslint.json]: I'm guessing this is just carried over from the version change, and that you didn't add it for some other reason?

Comment (v1zix) [reply types/react-leaflet/v1/tslint.json]: Correct, that's from here: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/09b9ef4ff0932791ff758e5c4c5baed0409f764f/types/react-leaflet/tslint.json#L4

**Pull request merged**